hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieSparkSqlWriter.scala

```diff
@@ -402,6 +389,40 @@ object HoodieSparkSqlWriter {
     }
   }
 
+  def deduceOperation(hoodieConfig: HoodieConfig, paramsWithoutDefaults : Map[String, String]): WriteOperationType = {
+    var operation = WriteOperationType.fromValue(hoodieConfig.getString(OPERATION))
+    // TODO clean up
+    // It does not make sense to allow upsert() operation if INSERT_DROP_DUPS is true
+    // Auto-correct the operation to "insert" if OPERATION is set to "upsert" wrongly
+    // or not set (in which case it will be set as "upsert" by parametersWithWriteDefaults()) .
+    if (hoodieConfig.getBoolean(INSERT_DROP_DUPS) &&
+      operation == WriteOperationType.UPSERT) {
+
+      log.warn(s"$UPSERT_OPERATION_OPT_VAL is not applicable " +
+        s"when $INSERT_DROP_DUPS is set to be true, " +
+        s"overriding the $OPERATION to be $INSERT_OPERATION_OPT_VAL")
+
+      operation = WriteOperationType.INSERT
+    }
+
+    // if no record key, no preCombine and no explicit partition path is set, we should treat it as append only workload
```

Comment:
not sure about this rule: shouldn't append-only just depends on having `preCombine` or not?

Reply:
to align on semantics, should these usage all result in bulk insert? 

```
df.write.option("table_name_config","tbl_name").format("hudi").save(basePath)
df.write.partitionBy("pt").option("table_name_config","tbl_name").format("hudi").save(basePath)
df.write.option("table_name_config","tbl_name").option("hoodie.datasource.write.partitionpath.field", "pt").format("hudi").save(basePath)
```

Reply:
yeah. had discussions w/ other hudi eng as well. 
here is the agreement. 
when none of record key, operation and preCombine is set, we will make bulk_insert as the operation type. in other cases (atleast in spark-ds), upsert will be  the operation type